File: spring/todo/src/main/java/teambydefault/todo/entity/Subtask.java
// //done with Kiro using Todo.java as a template

package teambydefault.todo.entity;

import java.time.LocalDateTime;
import java.util.UUID;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

@Entity
@Data
@NoArgsConstructor
@Table(name = "subtasks")
public class Subtask {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id")
    private UUID id;

    // Foreign key to to_dos table on task_id
    @ManyToOne
    @JoinColumn(name = "task_id", referencedColumnName = "task_id", nullable = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    private Todo todo;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "description")
    private String description;

    @Column(name = "due_date")
    private LocalDateTime dueDate;

    @Column(name = "is_completed")
    private Boolean isCompleted;
}

File: spring/todo/todo-front-end/src/app/interfaces/Subtask.ts
/*
  This interface mirrors the `Subtask` @Entity class in the Spring backend
  (src/main/java/teambydefault/todo/entity/Subtask.java).

  Field mapping from Java → TypeScript:
    UUID id               → string   (UUIDs are serialized as strings in JSON)
    String title          → string
    String description    → string
    LocalDateTime dueDate → string   (dates are serialized as ISO-8601 strings in JSON)
    boolean isCompleted   → boolean
    Todo todo             → { taskId: string }  (only the foreign key ID is needed on the frontend)
*/
export interface Subtask {
  id: string;
  title: string;
  description: string;
  dueDate: string;
  isCompleted: boolean;
  todo: { taskId: string };
}
